js-library changes

an-kumar opened the pull request:
This PR does a few things worth noting.

Fixes some issues with the babel process:
- removing `return` in `babel.js` allows mutliple files to be babelified.
- explicitly computing and passing in output paths to `babel.js` rather than using `basename` allows correct babelification of files in directories.

Making `babel` and `babelrc_tmpl` **not** private attributes:
- different people may have different babelrcs that they want to use. While babel does have some mechanisms for this such as nested babelrcs, I'd prefer if bazel was powerful enough to do this itself, since otherwise the babelification process is split across bazel and non-bazel. I'm also not sure if babel's mechanisms will actually work while using it in this way.
- if a user's babelrc is different enough from what's provided, the babel executable will need more dependencies (e.g, npm-installed babel plugins). We cannot pass them in as `data` to the `js_library` rule because of how rules_nodejs' module resolution work. Also, a current limitation of rules_nodejs with respect to npm dependencies is that they can only come from a single workspace, so that wouldn't work anyways. Instead, I just made `babel` a non-private attribute, and so the user can pass their own version of `babel` with the correct dependencies

Exposing `babel_executable` macro:
- This is just to make the above process easier: if a user needs extra dependencies for their babelrc, they can produce a viable `babel` executable very easily just using `babel_executable`.

Adding the ability to turn off sandboxing for the babel executable:
- On mac os, sandboxing is relatively costly for certain workloads, like the symlinking needed for nodejs_binary (e.g, the babel executable). Turning off sandboxing ends up being a reasonably large time saver on mac, since babel is run so often. So providing the ability to turn it off is nice.

Globegitter commented:
Thanks for the fixes and improvements - it does look mostly good. The reason I have not added customization yet was for simplicity to get a first version in and then think about how to allow for it. As now people potentially need to know about the `TMPL_` variables and then it could also break how it works with `ts_devserver` together.

But maybe it makes sense to have the full babel config overridable and then if we do come up with a better strategy we will have to bite the bullet and make it a breaking change. What do you think?

Globegitter commented:
Also do you think you could extend the examples that "test" the fixes you just made? Just so we have some way of seeing if we break anything later down the road.

an-kumar commented:
On customization -- I understand your concern; perhaps we shouldn't call it babelrc_tmpl, and instead just call it babelrc. The fact that it is implemented via a `TMPL_` scheme is sort of an implementation detail anyways, and fundamentally it is semantically just providing a babelrc. I do think that we should have the full babel config override though, primarily because the way the JS community uses babel is typically _because_ of it's customizability, so not allowing configuration sort of does not actually meet the bar of "adding babelification to bazel".

On tests -- yup, definitely, just wanted to make sure the changes are in line with what you expect. I'll get to it tonight or tomorrow probably.

Globegitter commented:
One thought I had is that we could potentially try and merge the user-specified babelrc with the default one - or another idea have two actions, one with the default config that always ensures that the basics work, i.e integration with `ts_devserver` and a second that is overridable with a user-provided config that is a pure babelrc without any templating stuff. Would still have to think a bit how that could integrate with the other rules though. Any thoughts?

an-kumar commented:
Hmm. Let me give it some thought.

First impression is that there's not really a great way of providing a "default config" that ensures the basics work, because different people use different features and require different babelification to arrive at something that would work with `ts_devserver`. In that sense the default config only ensures that the basics work **if** you only use those features that the default config correctly babelifies.

As for a second user-provided config, I'm just not sure that it is very idiomatic babel usage to sort of force users to have certain babelification parameters no matter what. I guess I'm not seeing a huge upside -- we can and should provide logical defaults to the rule, like what is there now, but ultimately I think it should be fully configurable, to be in line with the broader JS community paradigms.

[EDIT] On further inspection, I think I understand better your hesitation. Let me think about it a bit and see if I can think of anything.

an-kumar commented:
Okay, so I've thought about this a little. The use of templating right now provides the babelrc with the module name of the library being babelified. I understand your hesitation around making this a public attribute because then the templating would have to be explained to end users. I think we have a few options:

- Make the babelrc a public attribute and simply explain the templating scheme to end users. Many people may find the default babelrc suffices for them and never consider this.

- Make the babelrc a public attribute and **remove** templating entirely. If the user needs the module name in the babelrc, they can provide a different babelrc to every rule, with the module name baked in. The babelrc could, for example, be generated by a macro that does the templating to _produce_ the correct babelrc for the module, and then pass in that _generated_ file as the babelrc to `js_library`, which would look something like this:

```
def babel_library(name, srcs, module_name, etc...):
    some_rule_that_generates_templated_babelrc(name + "_babelrc", module_name)
   js_library(name, srcs, babelrc = name + "_babelrc")
```

I think you get the idea.

- Your suggestion around having a "default" babelrc, and allowing users to pass in a secondary babelrc.

I personally think the first two options are better than the third. I really don't like the idea of `js_library` pushing a babelification scheme onto the user without their consent. Generally speaking anything that modifies the users _code_ should probably be configurable IMO. Otherwise I think we'll run into a lot of issues where `js_library` doesn't work for some people because the forced babelification scheme messes up their code in some way.

What do you think?

Globegitter commented:
Another idea I just had is have a default babelrc that includes the templating (as it currently is) but when the user provides their custom babelrc do not do any templating at all. So per default it should work nicely out of the box and if people provide their custom template I think it is unlikely that they will need the module name or bin-dir path, so we do not have to document these as public attributes.

I might try and experiment around a bit with the different options to see what feels like the best option.

an-kumar commented:
I think that's a fine option as well.

We will still need to make both `babel` and `babelrc_tmpl` public. And with that approach, I think `babelrc_tmpl` should be renamed `babelrc.`

an-kumar commented:
I just updated the PR -- I left `_babelrc_tmpl` as a private attribute, but added a new public attribute `babelrc`. If `babelrc` is provided it is used, otherwise `_babelrc_tmpl` is templatized and used as before.

What do you think of that solution?

Globegitter commented:
@an-kumar Yeah I feel much happier about this solution. Let's go forward with that and we can see how it pans out once more people start using it. Are you happy to add a couple more examples or even e2e tests? Then I would be very happy to get this merged and that should also get the overall PR into a mergable state.

an-kumar commented:
Yep will add examples / tests soon now that we've settled on a solution 👍

Globegitter commented:
@an-kumar awesome, much apprecieated

an-kumar commented on internal/js_library/v2/js_library.bzl:
```diff
@@ -3,8 +3,15 @@
+
 def _write_config(ctx):
-  output = ctx.actions.declare_file(paths.join(ctx.file._babelrc_tmpl.dirname, "_" + ctx.file._babelrc_tmpl.basename))
+  output = ctx.actions.declare_file('%s.babelrc' % ctx.label.name)
```
i ran into issues where the templatized babelrc conflicted for different rules. Since we run _write_config for every rule anyways, this output name is just a unique one.

an-kumar commented:
I didn't see any tests for the new `js_library` stuff, just examples, so I added usage of the various changes here into the existing examples.

Let me know what you think.

jasongwartz commented:
@an-kumar thanks for the update! I'm also an infra developer (like Markus) from Ecosia.

Markus is on holiday for a couple weeks now and he's taken the lead on this, so I think it's best for him to take a look when he's back. Sorry for the delay!

Globegitter commented on internal/js_library/v2/package.json:
```diff
@@ -3,6 +3,8 @@
     "devDependencies": {
         "@babel/core": "^7.1.0",
         "@babel/preset-env": "^7.1.0",
+        "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
```
These 6.x plugins work with 7.x? Do they not have 7.x equivalents? Also adding it here means they get added for everyone, I think a solution would be nicer to have that as separate deps just for the example. But I think it is not supported to have deps from separate node_modules, is it?

an-kumar replied:
hmm, that's a good question. Honestly I just copied this from my code, but I guess it does work since it's used in an example.

Yeah, the issue is that you can't have deps from separate node_modules. I don't mind removing this though, I can probably find the 7.x equivalent. It's just for the example anyways.

Globegitter replied:
I guess that reduces the benefits of `babel_executable` a bit. Maybe it makes sense to have to provide all deps if it is being used? So then at least custom plugins etc can be provided. And yeah I would prefer not having these plugins in the default ones. I might just make the changes to get this merged.

an-kumar replied:
Yeah. It reduces the benefits of babel_executable a lot. Let's just remove babel_executable then, it makes it a little unclear what's happening. Users can just make their own babel using the `babel.js` file themselves if they want. I'll do this real quick.

Globegitter commented:
@an-kumar thanks for the examples and sorry for the delay in the response, as mentioned I just came back from my holiday yesterday.

an-kumar commented:
Hey I updated the PR. I removed `babel_executable` because as you pointed out while there is still a limitation on the node_modules it's really not that useful.

I also changed the `nodejs_binary` example directory to be it's own WORKSPACE, like some of the other examples there. That way, the flow babel plugins are in that directory's package.json, not distributed with the rules themselves.

an-kumar commented:
By the way, I'm realizing now I might have broken something on accident. My `_get_path` function, which I took from some other rules somewhere, was different than the logic in the `_trim` function you had in there, which I now realize was for the same goal (to ultimately get the path of the file relative to the package). I'm not super familiar with `ctx.bin_dir` and when it needs to be use. Could you confirm that I didn't break anything there?

I could just revert to your logic if you want. I didn't realize it was basically the same goal.

Globegitter commented:
Yeah that bit I was not too happy about. I'll take a look and see what's best. Thanks for all the changes seems we're getting close to merging :)

Globegitter commented:
The thing about the `ctx.bin_dir` is when you depend on another generated file. That has caused some confusion for me in the past, but if you have a generated file (so the output of a js_library or a genrule etc) these files are put into the `bin_dir` so if your js_library depends on another js_library those files will be in `bazel-out` (i.e. the bin_dir) rather than its real location.

I have created an issue about it here: https://github.com/bazelbuild/bazel/issues/5630
